### PR TITLE
Add mapViewModifier + Improve ShapeDataBuilder

### PR DESCRIPTION
**Pull request by hactar**
Currently with the lack of view modifiers, it is difficult to set up the mapView the way one may need it. For example, I see no way of setting `showUserLocation`, or changing the position of the compass. So I've created a `mapViewModifier` which allows you to set up the mapView with any of its existing properties, kind of like Introspect does for SwiftUI Views.

I do guess that this project does want to add specific modifiers for all properties at some time - `.compassPosition(.topLeft)` would be more SwiftUI than setting it up via this solution. But until all these modifiers exist this helper modifier would make this spm way more usable.

For more info see the in-code documentation.

### CODE REVIEW

File: Sources/MapLibreSwiftUI/MapView.swift [hunk 1]
```diff
@@ -9,6 +9,7 @@ public struct MapView: UIViewRepresentable {
 
     let styleSource: MapStyleSource
     let userLayers: [StyleLayerDefinition]
+    var mapViewModifier: ((MLNMapView) -> Void)?
```
Comment (ianthetechie) [inline Sources/MapLibreSwiftUI/MapView.swift:12]: In general, views in SwiftUI shouldn't contain mutable state outside the provided state frameworks.

I think this works for basic cases as-is, but I'm already getting a headache trying to sort out whether there is a failure mode in which an update isn't picked up.

Is it possible for your use case to make this `let` instead and add it to the constructor with a default value (`nil`)? You could probably also go one step further for ergonomics and write a modifier that returns a new `MapView` with this set to a new value (probably easiest to make a private "copy constructor" that takes the raw values rather than builders to facilitate)?

```suggestion
    let mapViewModifier: ((MLNMapView) -> Void)?
```

**Review by ianthetechie — COMMENTED** [review]
Thanks for this PR! Looks great in principle; I love this approach as it provides a completely generic "escape hatch" for now, unblocking devs until we decide the best conveniences to add later. (User location tracking is definitely one of them haha; just haven't figured out the best approach to adding these yet.)

Just one thing to discuss re: the way the `mapViewModifier` is set.

**Comment by ianthetechie** [comment]
Oh, and it would also be great if there were a preview showing this in action (basically copy your code to show the user location). These act as informal tests (and ensure we know where we're breaking APIs as it's still unstable) + showcases the functionality.

**Review by Archdoog — COMMENTED** [review]
I absolute see the necessity for something like this given how much of the maplibre library we would need to support to truly capture the full scope of the MapView. 

However, I also see this making the process of building the right solutions for a specific problem more challenging. Let's take the example where someone's directly manipulating some camera parameter on the MLNMapView then moments later the MapViewCamera is being processed off of state. Basically, we're now allowing untracked state manipulation potentially leading to a host of issues like diverging from the MapViewCamera state or the view redrawing over and over.

I'd almost propose we have a `FreeformMapView` or an `UnsafeMapView` with this. It would be a useful tool to understand a feature we need a properly scoped modifier for, etc.

**Comment by hactar** [comment]
I'm im full agreement here that opening this "backdoor" is a bypass, not a solution - but one I'd deem necessary with the current state of things - I'll modify the PR based on your feedback and try to give it a better name so that if anyone uses it they are aware that they may brake state management if used incorrectly.

Regarding the one time callback, while that might be useful for fire and forget set up modifiers (turn off logo for example), it won't work for many basic use cases, for example:

Showing a map with `.showsUserLocation = true` triggers a location permission request. In my use case I don't want that pop up appearing over the map at launch, I want to trigger it when a user taps the "locate me" button which launches the permission flow. So when the map view first shows, `.showUserLocation` needs to be false. Only once authorization has been granted do I want to flip `.showUserLocation` to true. My code for this looks like this:

```Swift
    @State var showUserLocation = false
// ...
       MapView(styleURL: styleURL, camera: $camera) {
       // layers setup...
            
        }
            .mapViewModifier { mapView in
                mapView.showsUserLocation = showUserLocation
                mapView.compassViewPosition = .topLeft
            }
            .task {
                for await event in await locationManager.startMonitoringAuthorization() {
                    if event.authorizationStatus == .authorizedWhenInUse {
                        showUserLocation = true
                    } else {
                        showUserLocation = false
                    }
                }
            }
```

And this code wouldn't work if we only used the `make` portion of UIViewRepresentable... Unless you see some better way of doing this that I'm missing.

File: Sources/MapLibreSwiftUI/MapView.swift (hunk 1 above)
Comment (hactar) [reply Sources/MapLibreSwiftUI/MapView.swift:12]: So my thought here was that as this isn't defined as `public`, the only way to modify this is from outside is via the defined modifier - as this is all a Struct I do not see how an update could not be picked up? By making it a var we let swift handle the copying for us if needed instead of having to maintain the private copy constructor and modifier every time we modify the parameters of MapView. Instead I would rename the modifier to unsafeMapViewModifier and add documentation to explain that this is an escape hatch for now and "here be dragons".

Alternatively we not make this a modifier at all, but just part of `init` as you suggest with a default of nil - I'm personally just not a fan of double closures - but they are very common in SwiftUI, like for `Button`, so in this case I guess that would be fine too.

I will update the PR to unsafeMapViewModifier first so we see how it feels, but happy to change it to a non modifier if it doesn't feel good enough.

**Comment by hactar** [comment]
Alright - applied changes, updated documentation, added an example and added a test for ShapeSource. Personally I think adding it to "init" would make it too prominent - I don't see a way of making this a "permanent" solution because the more modifiers are added, the more likely someone will try to override something that they should be controlling via a DSL modifier that already exists. But happy to remove the modifier and replace it with an init closure instead if you guys feel thats more in line with what you have planned.

File: Sources/MapLibreSwiftUI/MapView.swift (hunk 1 above)
Comment (ianthetechie) [reply Sources/MapLibreSwiftUI/MapView.swift]: Ignore me; that was a dum comment. I think you're correct.

**Comment by ianthetechie** [comment]
@hactar thanks for your thorough comments!

@Archdoog and I discussed this today and we agree that...

1. You have a valid use case which we can't yet think of a better way to address
2. It isn't going to be detrimental in our current "real-world use case" (building a navigation map view in [Ferrostar](https://github.com/stadiamaps/ferrostar)) as we just won't be exposing this detail to library users there anyways, and
3. Marking it clearly unsafe and explaining the pitfalls is more of a courtesy to new users than SwiftUI ever gave us 😂 

So long story short, we're cool with merging this. I just have a minor suggestion to the docs to make things more explicit as to why it's potentially unsafe.

**Comment by hactar** [comment]
Done - thanks :)
